doc/man/man3/Boss.3
.TH "Boss" 3 "Version 1.0" "Zaxxon" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Boss \- The boss entity you fight at the end of the level loop\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <Boss\&.h>\fP
.PP
Inherits \fBEntity\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBBoss\fP (sf::Vector3f, \fBEntity\fP *, sf::Texture *, sf::Texture *)"
.br
.RI "Prepare the boss variables\&. "
.ti -1c
.RI "\fB~Boss\fP () override"
.br
.RI "Clean up memory related to the boss\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &, float gameSpeed) override"
.br
.RI "Run the logic for the boss each frame and then draw it to the screen\&. "
.ti -1c
.RI "void \fBhit\fP ()"
.br
.RI "Run visual effects and code for when you land a hit on the boss\&. "
.ti -1c
.RI "\fBBossBullet\fP * \fBgetMissile\fP () const"
.br
.RI "Get the boss missile separately\&. "
.ti -1c
.RI "bool \fBmissileCreated\fP () const"
.br
.RI "Check if the boss missile exists\&. "
.ti -1c
.RI "bool \fBisDestroyed\fP () const"
.br
.RI "Check if the boss has been defeated\&. "
.in -1c

Public Member Functions inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "\fBEntity\fP ()"
.br
.RI "Setup \fBEntity\fP class variables\&. "
.ti -1c
.RI "virtual \fB~Entity\fP ()=0"
.br
.RI "Clean up memory related to \fBEntity\fP class\&. "
.ti -1c
.RI "virtual void \fBkill\fP (\fBAnimation::Anim\fP \fBanimation\fP=\fBAnimation::CHARACTER_DEATH\fP)"
.br
.RI "Run death animation code for the killed entity\&. "
.ti -1c
.RI "int \fBgetAnimationState\fP ()"
.br
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f pos)"
.br
.ti -1c
.RI "sf::Vector3f \fBgetPos\fP ()"
.br
.ti -1c
.RI "sf::FloatRect \fBgetBounds\fP ()"
.br
.ti -1c
.RI "sf::Vector2f \fBgetSpritePos\fP ()"
.br
.RI "Get the position of the related sprite for this entity\&. "
.in -1c
.SS "Additional Inherited Members"


Protected Attributes inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "sf::Texture * \fBspriteSheet\fP"
.br
.ti -1c
.RI "sf::Sprite * \fBsprite\fP"
.br
.ti -1c
.RI "\fBAnimation\fP \fBanimations\fP"
.br
.ti -1c
.RI "bool \fBdead\fP = false"
.br
.ti -1c
.RI "sf::Text \fBdebugText\fP"
.br
.ti -1c
.RI "sf::Font \fBdebugFont\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
The boss entity you fight at the end of the level loop\&. 
.PP
Definition at line \fB11\fP of file \fBBoss\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Boss::Boss (sf::Vector3f start, \fBEntity\fP * target, sf::Texture * bossSheet, sf::Texture * spriteSheet)"

.PP
Prepare the boss variables\&. 
.PP
\fBParameters\fP
.RS 4
\fIstart\fP 
.br
\fItarget\fP 
.br
\fIbossSheet\fP 
.br
\fIspriteSheet\fP 
.RE
.PP

.PP
Definition at line \fB11\fP of file \fBBoss\&.cpp\fP\&.
.SS "Boss::~Boss ()\fR [override]\fP"

.PP
Clean up memory related to the boss\&. 
.br
 
.PP
Definition at line \fB40\fP of file \fBBoss\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBBossBullet\fP * Boss::getMissile () const"

.PP
Get the boss missile separately\&. 
.PP
\fBReturns\fP
.RS 4
The missile as a BossBullet*
.RE
.PP

.PP
Definition at line \fB161\fP of file \fBBoss\&.cpp\fP\&.
.SS "void Boss::hit ()"

.PP
Run visual effects and code for when you land a hit on the boss\&. 
.PP
Definition at line \fB139\fP of file \fBBoss\&.cpp\fP\&.
.SS "bool Boss::isDestroyed () const"

.PP
Check if the boss has been defeated\&. 
.PP
\fBReturns\fP
.RS 4
A boolean
.RE
.PP

.PP
Definition at line \fB181\fP of file \fBBoss\&.cpp\fP\&.
.SS "bool Boss::missileCreated () const"

.PP
Check if the boss missile exists\&. 
.PP
\fBReturns\fP
.RS 4
A boolean
.RE
.PP

.PP
Definition at line \fB171\fP of file \fBBoss\&.cpp\fP\&.
.SS "void Boss::update (sf::RenderWindow & window, float gameSpeed)\fR [override]\fP, \fR [virtual]\fP"

.PP
Run the logic for the boss each frame and then draw it to the screen\&. 
.PP
\fBParameters\fP
.RS 4
\fIwindow\fP 
.RE
.PP

.PP
Implements \fBEntity\fP\&.
.PP
Definition at line \fB50\fP of file \fBBoss\&.cpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Zaxxon from the source code\&.


src/Entity/Boss/Boss.cpp
#include "Boss.h"


/// <summary>
/// Prepare the boss variables
/// </summary>
/// <param name="start"></param>
/// <param name="target"></param>
/// <param name="bossSheet"></param>
/// <param name="spriteSheet"></param>
Boss::Boss(sf::Vector3f start, Entity* target, sf::Texture* bossSheet,
           sf::Texture* spriteSheet) : Entity()
{
	setPos(start);

	sprite->setPosition(translateTo2d(start));
	sprite->setTexture(*bossSheet);
	sprite->setTextureRect(sf::IntRect(0, 0, 58, 75));
	sprite->setOrigin(
		sf::Vector2f(0, sprite->getGlobalBounds().height * (3.f / 4.f)));

	this->target = target;

	this->spriteSheet = spriteSheet;

	movementInt.restart();

	//defines random movement, goes to two seperate points
	targetXPoints[0] = (rand() % 100) * -1.f;
	targetXPoints[1] = (abs((rand() - 528) * 72) % 100) * -1.f;

	//timer for invincibility frames
	invFrames.restart();
}


/// <summary>
/// Clean up memory related to the boss
/// </summary> 
Boss::~Boss()
{
	delete missile;
}


/// <summary>
/// Run the logic for the boss each frame and then draw it to the screen
/// </summary>
/// <param name="window"></param>
void Boss::update(sf::RenderWindow& window, float gameSpeed)
{
	if (movementInt.getElapsedTime().asMilliseconds() >= 100)
	{
		//moves to the target position every 100 millisecond to mimic gittery 
		//movment of the game
		movementInt.restart();

		if (targetXPoints[stages] - getPos().x > 3)
			setPos(sf::Vector3f(getPos().x + 3, getPos().y, getPos().z));

		if (targetXPoints[stages] - getPos().x < 3)
			setPos(sf::Vector3f(getPos().x - 3, getPos().y, getPos().z));

		if (abs(getPos().z - target->getPos().z) < 250 && stages == 0)
			stages++;

		//if the boss is far enough away from the player continue moving it
		if (abs(getPos().z - target->getPos().z) > 150)
		{
			if (stages < 2)
				setPos(sf::Vector3f(getPos().x, getPos().y, getPos().z + 3));
			else
				setPos(sf::Vector3f(getPos().x, getPos().y, getPos().z - 7));
		}
		//else stop it and shoot the missile
		else
		{
			sprite->setTextureRect(sf::IntRect(58, 0, 58, 75));
			stages++;
			setPos(sf::Vector3f(getPos().x, getPos().y, getPos().z - 7));

			bulletCreated = true;
			//gives the missile the position of the player
			missile = new BossBullet(
				sf::Vector3f(getPos().x - 33, getPos().y - 19, getPos().z),
				target, spriteSheet);
			missile->damage(hits);
		}
	}

	//detects the hits to the missile and adds the shaking animation
	if (movementInt.getElapsedTime().asMilliseconds() % 25)
	{
		if (hitCount < 10)
		{
			hitCount++;
			setPos(sf::Vector3f(getPos().x + ((rand() % 100) / 50.f) - 1,
			                    getPos().y + ((rand() % 100) / 50.f) - 1,
			                    getPos().z));
		}
		else if (hitCount == 10)
		{
			hitCount++;
			sprite->setColor(sf::Color(255, 255, 255));
		}
	}

	sprite->setPosition(translateTo2d(getPos()));

	//if it retreats all the way it has been defeated
	if (getPos().z <= -4000 && stages >= 2)
	{
		destroyed = true;
		stages++;
	}

	targetXPoints[2] = getPos().x;

	window.draw(*sprite);

	//if the missile dies delete it and record that action
	if (stages == 2 && missile != nullptr)
	{
		if (missile->isHit())
		{
			delete missile;
			missile = nullptr;
			bulletCreated = false;
		}
		else
			missile->update(window, gameSpeed);
	}
}


/// <summary>
/// Run visual effects and code for when you land a hit on the boss.
/// </summary>
void Boss::hit()
{
	//damages the boss and starts the timer of invincibility
	if (invFrames.getElapsedTime().asMilliseconds() >= 200)
	{
		invFrames.restart();
		hitCount = 0;
		sprite->setColor(sf::Color(225, 100, 100));
		hits++;
		if (hits >= 10)
		{
			sprite->setTextureRect(sf::IntRect(58, 0, 58, 75));
			stages = 3;
		}
	}
}


/// <summary>
/// Get the boss missile separately.
/// </summary>
/// <returns>The missile as a BossBullet*</returns>
BossBullet* Boss::getMissile() const
{
	return missile;
}


/// <summary>
/// Check if the boss missile exists.
/// </summary>
/// <returns>A boolean</returns>
bool Boss::missileCreated() const
{
	return bulletCreated;
}


/// <summary>
/// Check if the boss has been defeated.
/// </summary>
/// <returns>A boolean</returns>
bool Boss::isDestroyed() const
{
	return destroyed;
}
